Support navigating cross-site iframe to mainFrame domain

chirags27 opened the pull request:
#### 4ca041a13dfafa4a2fbe427511dd2705969b1be8
<pre>
Support navigating cross-site iframe to mainFrame domain
<a href="https://bugs.webkit.org/show_bug.cgi?id=256339">https://bugs.webkit.org/show_bug.cgi?id=256339</a>

Reviewed by Alex Christensen.

This change adds support so that an iframe can navigate back to the
mainFrame domain, and so that the load happens in the mainFrame process.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::addRootFrame):
* Source/WebCore/page/Page.h:
(WebCore::Page::addRootFrame): Deleted.
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::provisionalFrame):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForResponse):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::transitionToLocal):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::SandboxExtensionTracker::beginLoad):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/263747@main">https://commits.webkit.org/263747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6011356407a861c6a3065277e48a39ed0ba28e95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5645 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5988 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7196 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5630 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5644 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6021 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5770 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7737 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5749 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5797 "2 new passes 10 flakes 1 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5056 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7238 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3274 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5082 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12294 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5143 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5158 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6986 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5593 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4573 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5041 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9152 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/648 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5402 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

chirags27 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3914,7 +3915,7 @@ void WebPageProxy::receivedNavigationPolicyDecision(PolicyAction policyAction, A
     }
 
-    Ref<WebProcessProxy> sourceProcess = process();
+    Ref<WebProcessProxy> sourceProcess = sourceProcessForNavigation;
```
I think this change is leading to failure in `SiteIsolation.ChildNavigatingToNewDomain`:

```
ASSERTION FAILED: !m_messageReceivers.contains(std::make_pair(messageReceiverName, destinationID))
/Volumes/Data/safari-webkit/OpenSource/Source/WebKit/Platform/IPC/MessageReceiverMap.cpp(53) : void IPC::MessageReceiverMap::addMessageReceiver(ReceiverName, uint64_t, MessageReceiver &)
1   0x10df13d08 WTFCrash
2   0x12c138814 WTFCrashWithInfo(int, char const*, char const*, int)
3   0x12f14fc70 IPC::MessageReceiverMap::addMessageReceiver(IPC::ReceiverName, unsigned long long, IPC::MessageReceiver&)
4   0x12dcafca0 WebKit::AuxiliaryProcessProxy::addMessageReceiver(IPC::ReceiverName, unsigned long long, IPC::MessageReceiver&)
5   0x12d7bb1bc WebKit::AuxiliaryProcessProxy::addMessageReceiver(IPC::ReceiverName, WTF::ObjectIdentifierGenericBase const&, IPC::MessageReceiver&)
6   0x12dd6093c WebKit::SubframePageProxy::SubframePageProxy(WebKit::WebPageProxy&, WebKit::WebProcessProxy&, bool)
7   0x12dd60b48 WebKit::SubframePageProxy::SubframePageProxy(WebKit::WebPageProxy&, WebKit::WebProcessProxy&, bool)
8   0x12dfb7be0 WTF::UniqueRef<WebKit::SubframePageProxy> WTF::makeUniqueRefWithoutFastMallocCheck<WebKit::SubframePageProxy, WebKit::WebPageProxy&, WTF::Ref<WebKit::WebProcessProxy, WTF::RawPtrTraits<WebKit::WebProcessProxy>>&, bool>(WebKit::WebPageProxy&, WTF::Ref<WebKit::WebProcessProxy, WTF::RawPtrTraits<WebKit::WebProcessProxy>>&, bool&&)
```

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1956,7 +1956,6 @@ void Page::finalizeRenderingUpdate(OptionSet<FinalizeRenderingUpdateFlags> flags
 void Page::finalizeRenderingUpdateForRootFrame(LocalFrame& rootFrame, OptionSet<FinalizeRenderingUpdateFlags> flags)
 {
     LOG(EventLoop, "Page %p finalizeRenderingUpdate()", this);
-
```
This change is not relevant to this PR.

achristensen07 commented on Source/WebCore/page/Page.h:
```diff
@@ -1041,7 +1041,7 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
 
     const WeakHashSet<LocalFrame>& rootFrames() const { return m_rootFrames; }
-    void addRootFrame(LocalFrame& frame) { m_rootFrames.add(frame); }
+    void addRootFrame(LocalFrame& frame) { ASSERT(frame.isRootFrame()); m_rootFrames.add(frame); }
```
This should be multiple lines.  Probably move it to Page.cpp.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -77,6 +77,7 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
 
     // FIXME: This gives too much cookie access. This should be removed after putting the entire frame tree in all web processes.
     auto giveAllCookieAccess = LoadedWebArchive::Yes;
+
```
This change is not relevant to this PR.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6432,8 +6436,12 @@ void WebPageProxy::decidePolicyForNewWindowAction(FrameIdentifier frameID, Frame
 void WebPageProxy::decidePolicyForResponse(IPC::Connection& connection, FrameIdentifier frameID, FrameInfoData&& frameInfo, PolicyCheckIdentifier identifier, uint64_t navigationID, const ResourceResponse& response, const ResourceRequest& request, bool canShowMIMEType, const String& downloadAttribute, uint64_t listenerID)
 {
     RefPtr frame = WebFrameProxy::webFrame(frameID);
+    // If a provisional frame loading is going on, we need to use the process from that
```
This comment isn't helpful.

chirags27 commented on Source/WebKit/UIProcess/SubframePageProxy.cpp:
```diff
@@ -104,6 +104,7 @@ void SubframePageProxy::decidePolicyForResponse(WebCore::FrameIdentifier frameID
 
 void SubframePageProxy::didCommitLoadForFrame(WebCore::FrameIdentifier frameID, FrameInfoData&& frameInfo, WebCore::ResourceRequest&& request, uint64_t navigationID, const String& mimeType, bool frameHasCustomContentProvider, WebCore::FrameLoadType frameLoadType, const WebCore::CertificateInfo& certificateInfo, bool usedLegacyTLS, bool privateRelayed, bool containsPluginDocument, WebCore::HasInse ...
 {
+    m_process->didCommitProvisionalLoad();
```
This is needed to that we don't end up using the same process because of:

```
    if (!sourceProcess->hasCommittedAnyProvisionalLoads()) {
        tryPrewarmWithDomainInformation(sourceProcess, targetRegistrableDomain);
        return { WTFMove(sourceProcess), nullptr, "Process has not yet committed any provisional loads"_s };
    }
```

now that we pass the right source process to `WebProcessPool::processForNavigation`

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6433,7 +6437,10 @@ void WebPageProxy::decidePolicyForResponse(IPC::Connection& connection, FrameIde
     MESSAGE_CHECK_BASE(frame, &connection);
-    decidePolicyForResponseShared(Ref { frame->process() }, internals().webPageID, frameID, WTFMove(frameInfo), identifier, navigationID, response, request, canShowMIMEType, downloadAttribute, listenerID);
+    if (auto provisionalFrameProxy = frame->provisionalFrame())
+        decidePolicyForResponseShared(Ref { provisionalFrameProxy->process() }, internals().webPageID, frameID, WTFMove(frameInfo), identifier, navigationID, response, request, canShowMIMEType, downloadAttribute, listenerID);
```
This is a lot of duplicate code.  Could we do something like this instead?
auto* provisionalFrame = frame->provisionalFrame();
Ref sourceProcess { provisionalFrame ? provisionalFrame->process() : frame->process() };
decidePolicyForResponseShared(sourceProcess, ...)

chirags27 committed:
Support navigating cross-site iframe to mainFrame domain
https://bugs.webkit.org/show_bug.cgi?id=256339

Reviewed by Alex Christensen.

This change adds support so that an iframe can navigate back to the
mainFrame domain, and so that the load happens in the mainFrame process.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::addRootFrame):
* Source/WebCore/page/Page.h:
(WebCore::Page::addRootFrame): Deleted.
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::provisionalFrame):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForResponse):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::transitionToLocal):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::SandboxExtensionTracker::beginLoad):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/263747@main
